R/IDateTime.R

```diff
@@ -5,9 +5,17 @@
 
 as.IDate <- function(x, ...) UseMethod("as.IDate")
 
-as.IDate.default <- function(x, ..., tz = attr(x, "tzone")) {
+i.IDate = NULL
+as.IDate.default <- function(x, ..., tz = attr(x, "tzone"), use_merge = 'auto') {
   if (is.null(tz)) tz = "UTC"
-  as.IDate(as.Date(x, tz = tz, ...))
+  if (isTRUE(use_merge) || (use_merge == 'auto' && length(x) >= 1000)) {
+    DT = data.table(x)
+    # shut off use_merge to prevent recursion
+    unique(DT)[ , 'IDate' := as.IDate(x, tz = tz, ..., use_merge = FALSE)
```

Comment:
Could you elaborate? Not quite following the reasoning... my thinking was `unique.data.table` would be more efficient.

Reply:
It will be more efficient than data.frame method, but not sure about comparing to vector. R uses data.table's order (not the parallel one, but previous one, still super fast). The fastest way might be to use `forder(retGrp=T)` to then subset to unique value.

Reply:
when using `unique.data.table` we have internal `funique` function in setops.R

Reply:
I didn't think about doing merge with forder, just unique, but you are correct. We can emulate merge with forder, in such simple case of single column self join it should be relatively faster.

Reply:
OK, not immediate to me how to do it with just `forderv`... here's what I came up with and it's horribly inefficient:

```
    o = forderv(DT, retGrp = TRUE)
    s = attr(o, 'starts')
    n = length(x)
    out = integer(n)
    lookup = as.IDate(as.Date(x[o][s], tz = tz, ...))
    for (ii in seq_along(s)) {
      end_idx = if (ii == length(s)) n else s[ii + 1L] - 1L
      out[o][s[ii]:end_idx] = lookup[ii]
    }
    setattr(out, 'class', c('IDate', 'Date'))

Reply:
which part is inefficient? loop? `src/vecseq.c` might be good solution for that
other ideas:
- `x` is reordered according to `o` and then subset according to `s`. Shouldn't that be better to first subset and then reorder?
```r
cc(F)
x = c(2L,4L,1L,2L,3L,1L,4L)
o = forderv(data.table(x), "x", retGrp=TRUE)
s = attr(o, 'starts')
x[o][s]
x[o[s]]
```
- instead of `as.IDate(as.Date(` it should be better to handle that in `as.IDate` only, if possible

Reply:
Here's an approach with ```forderv```:

```r
IDate_character = function (x) {
  o__ = data.table:::forderv(list(x), 1L, retGrp=TRUE)
  if (attr(o__, "maxgrpn") == 1L) { ## lookup will not help us - let's just make a Date
    ans = as.integer(as.Date(x))
  } else {
    f__ = attr(o__, 'starts')
    len__ = data.table:::uniqlengths(f__, length(x))
    
    if (!length(o__)) {
      ans = rep.int(as.integer(as.Date(x[f__])), len__)
    } else {
      ans = integer(length(x))
      ans[o__] = rep.int(as.integer(as.Date(x[o__[f__]])), len__)
      }
  }
  class(ans) = c("IDate", "Date")
  return(ans)
}
```
Initial testing is promising - more performant than a lookup self-join and results seem to align with ```as.IDate```. But I will admit it does not look as clean as the join method.

Reply:
Great work! for our internal code I think this sort of usage of `forderv` is common enough that we don't need to worry too much about "cleanliness" as there are a few other examples too.

A benchmark would be nice; in any case feel free to update/mege directly into this branch to be sure you're getting the commit credit :smiley:

Reply:
There are 3 options we are discussing.

1. Make a lookup table based on unique dates
2. Use an existing lookup table that would invisibly take up memory (but be fast)
3. Use ```forderv``` directly

Additionally, there are two flavors, unkeyed and keyed. For the most part, option 1 is the fastest although once we get to 1e8 rows, the ```forderv``` option surprisingly starts to outpace it. Note, option #2 (which is proposed in the FR) is faster than my proposal with keyed and 1e5 rows. 

If we agree that this is limited to ```as.IDate.character()```, I will make the changes. 

Unkeyed:
![image](https://user-images.githubusercontent.com/57992489/93727259-5f3a4600-fb88-11ea-9dd7-4b89212ec054.png)

Keyed:
![image](https://user-images.githubusercontent.com/57992489/93727502-6f065a00-fb89-11ea-9081-6e1a79205af9.png)

Code used:
```r
library(data.table)

IDate_character = function (x) {
  o__ = data.table:::forderv(list(x), 1L, retGrp=TRUE)
  if (attr(o__, "maxgrpn", exact = TRUE) == 1L) { ## lookup will not help us - let's just make a Date
    ans = as.integer(as.Date(x))
  } else {
    f__ = attr(o__, 'starts', exact = TRUE)
    len__ = data.table:::uniqlengths(f__, length(x))
    
    if (!length(o__)) {
      ans = rep.int(as.integer(as.Date(x[f__])), len__)
    } else {
      ans = integer(length(x))
      ans[o__] = rep.int(as.integer(as.Date(x[o__[f__]])), len__)
    }
  }
  class(ans) = c("IDate", "Date")
  return(ans)
}

IDate_lookup = function(x) {
  tz = "UTC"
  DT = list(input = x)
  setDT(DT)
  
  lookup = unique(DT)
  lookup[ , 'IDate' := as.IDate(as.Date(input, tz = tz))]
  lookup[DT, on = 'input']$IDate
}

IDate_existing_lookup = function(x) {
  tz = "UTC"
  DT = list(date_char = x)
  setDT(DT)
  
  date_lookup[DT, on = 'date_char']$date_int
}

dates = seq.Date(as.Date('1900-01-01'),
                 as.Date('2099-12-31'), by = 'day')
date_lookup = data.table(
  date_char = as.character(dates),
  date_int = as.IDate(dates),
  key = 'date_char'
)

set.seed(3082)
NN = 1e7
smp_dt = copy(date_lookup[sample(.N, NN, TRUE)])
setkey(smp_dt, date_char)

plot(bench::press(
  NN = c(10^(1:8)),
  {
    set.seed(3082)
    smp_dt = copy(date_lookup[sample(.N, NN, TRUE)])
    setkey(smp_dt, date_char) ## comment out to do unkeyed
    bench::mark(
      IDate_character(smp_dt$date_char),
      IDate_lookup(smp_dt$date_char),
      IDate_existing_lookup(smp_dt$date_char),
      # as.IDate(smp_dt$date_char)
      , min_iterations = 2L
    )
  }
))
```